### PR TITLE
Add support for UDS and existing Channels

### LINKED ISSUE #259: Support establishing connections over existing NIO channels
It is useful to be able to ask MySQLNIO to initiate a connection using a preexisting NIO channel. The most obvious use case is SSH tunneling. This should mostly be a matter of simply exposing the appropriate API as public.

See also vapor/mysql-nio#76.

/cc @fabianfett

### LINKED ISSUE #268: The new connect syntax appears to be missing an option for unix domain sockets
The new connect syntax calls one of the following swift-nio connect functions:
`func connect(host: String, port: Int) -> EventLoopFuture<Channel>`
`func connect(to address: SocketAddress) -> EventLoopFuture<Channel>`

it dosen't call the third:
`func connect(unixDomainSocketPath: String) -> EventLoopFuture<Channel>`

### PR DESCRIPTION
**Pull request by gwynne**
Extends `PostgresConnection` to enable connecting to Postgres servers over UNIX domain sockets or using a preexisting `Channel`.

In the process, most of the existing structure of `PostgresConnection.Configuration` has been deprecated in favor of new APIs with more options and cleaner design. The deprecation messages provide instructions on migrating existing code. List of deprecations and their replacements (all listed APIs should be assumed to have a `PostgresConnection.` prefix):

- `Configuration.Connection` and `.connection` -> `Configuration.host`, `.port`, `.unixDomainSocket`, and `.establishedChannel` properties, and respective initializers.
- `Configuration.Authentication` and `.authentication` -> `Configuration.username`, `.password`, and `.database` properties.
- New `Configuration.Options` structure and `.options` property, containing the `connectTimeout` and `requireBackendKeyData` properties previously found on `Configuration.Connection`.

_Note:_ The integration tests for UNIX domain sockets will be skipped unless `POSTGRES_SOCKET` is set to a socket path in the test environment. See the updated `test.yml` for examples of how to do so correctly.

Closes #259
Closes #268

### CODE REVIEW

**Comment by gwynne** [comment]
This PR is intended to subsume both #302 and #314, lmk if there's anything missing.

**Comment by 0xTim** [comment]
Resolves #259

**Comment by gwynne** [comment]
CI failure on `nightly-main` is tracked by https://github.com/apple/swift/issues/65064

**Comment by gwynne** [comment]
Also resolves #268

**Comment by gwynne** [comment]
Resolves #268

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -27,7 +31,7 @@ jobs:
       - name: Run unit tests with code coverage and Thread Sanitizer
         run: swift test --enable-test-discovery --filter=^PostgresNIOTests --sanitize=thread --enable-code-coverage
       - name: Submit coverage report to Codecov.io
-        if: "!contains(matrix.container, 'nightly')"
+        if: ${{ !contains(matrix.container, '5.8') }}
```
Comment (fabianfett) [inline .github/workflows/test.yml:34]: why did we change this to `5.8`?

Comment (MahdiBM) [reply .github/workflows/test.yml:34]: The reason we used 'nightly' was that Codecov seemed to have problems with nightly builds. 
We can either try to see if Codecov now works fine with nightly builds, or let it stay disabled to make sure it doesn't fail the CI. 
I think Gwynne thought now that 5.8 is released, we can change that to '5.8', but if i were to guess i would say the problems we were having were nightly related, not 5.8's fault.

File: Sources/PostgresNIO/New/Data/UUID+PostgresCodable.swift [hunk 2]
```diff
@@ -1,4 +1,5 @@
 import NIOCore
+import NIOFoundationCompat
```
Comment (fabianfett) [inline Sources/PostgresNIO/New/Data/UUID+PostgresCodable.swift:2]: why did that happen?

File: Tests/IntegrationTests/PostgresNIOTests.swift [hunk 3]
```diff
@@ -31,6 +31,25 @@ final class PostgresNIOTests: XCTestCase {
...
+    
+    func testConnectUDSAndClose() throws {
+        try XCTSkipUnless(env("POSTGRES_SOCKET") != nil)
+        var conn: PostgresConnection?
+        XCTAssertNoThrow(conn = try PostgresConnection.testUDS(on: eventLoop).wait())
+        XCTAssertNoThrow(try conn?.close().wait())
+    }
+    
+    func testConnectEstablishedChannelAndClose() throws {
+        let channel = try ClientBootstrap(group: self.group).connect(to: PostgresConnection.address()).wait()
```
Comment (fabianfett) [inline Tests/IntegrationTests/PostgresNIOTests.swift:43]: do we need
```
        try XCTSkipUnless(env("POSTGRES_SOCKET") != nil)
```
here as well?

File: .github/workflows/test.yml (hunk 1 above)
Comment (MahdiBM) [reply .github/workflows/test.yml:34]: I see Gwynne is tracking the Codecov issue already, based on her comment above. Then i guess the options are to either remove that 'if' or change it back.

**Review by 0xTim — APPROVED** [review]
LGTM - only thing to mention that I couldn't quite tell is that it looks like all the tests have been updated to use the new APIs. We should probably have a test just to ensure the deprecated stuff still works

**Review by MahdiBM — APPROVED** [review]

File: Sources/PostgresNIO/Connection/PostgresConnection.swift [hunk 4]
```diff
@@ -68,31 +68,209 @@ public final class PostgresConnection {
...
             ///
-            /// - Default: 5432
-            public var port: Int
+            /// - Parameters:
+            ///   - channel: The ``NIOCore/Channel`` to use. The channel must already be active and
+            ///     connected to an endpoint.
+            ///   - tlsHostname: See ``tlsHostname``.
+            ///   - requireBackendKeyData: See ``requireBackendKeyData``.
+            public static func establishedChannel(
+                channel: Channel,
```
Comment (fabianfett) [inline Sources/PostgresNIO/Connection/PostgresConnection.swift:143]: if we use `establishedChannel` then we should drop the `channel` parameter name

File: Sources/PostgresNIO/Connection/PostgresConnection.swift [hunk 5]
```diff
@@ -269,8 +447,16 @@ public final class PostgresConnection {
...
-            case .unresolved(let host, let port):
+            case .unresolvedTCP(let host, let port):
                 connectFuture = bootstrap.connect(host: host, port: port)
+            case .unresolvedUDS(let path, _):
+                connectFuture = bootstrap.connect(unixDomainSocketPath: path)
+            case .bootstrapped(let channel, _):
+                guard channel.isActive else {
+                    return eventLoop.makeFailedFuture(PSQLError.channel(underlying: ChannelError.alreadyClosed))
+                }
+                // TODO: Are there drawbacks to creating a bootstrap we don't end up using?
```
Comment (fabianfett) [inline Sources/PostgresNIO/Connection/PostgresConnection.swift:458]: I think we are fine here. The bootstrap is minimal anyway

File: .github/workflows/test.yml (hunk 1 above)
Comment (gwynne) [reply .github/workflows/test.yml:34]: The issue was that 5.8 changed the behavior of the `--show-codecov-path` flag. This was fixed, but the fix didn't make it in until 5.9. See https://github.com/apple/swift-package-manager/issues/5853 for details.

File: Sources/PostgresNIO/New/Data/UUID+PostgresCodable.swift (hunk 2 above)
Comment (gwynne) [reply Sources/PostgresNIO/New/Data/UUID+PostgresCodable.swift:2]: It's required for `ByteBuffer.writeUUIDBytes(_:)` and `.readUUIDBytes()`. The need for the import becomes explicit when building in `BUILDING_DOCC` mode (which basically removes `Exports.swift` and thus all uses of `@_exported`). This was our workaround for the issues DocC has with re-exported imports - `@_documentation(visibility:)` didn't come along until later.

File: Tests/IntegrationTests/PostgresNIOTests.swift (hunk 3 above)
Comment (gwynne) [reply Tests/IntegrationTests/PostgresNIOTests.swift:43]: That's only for the UDS tests; for the established channel test I just mimic the channel bootstrap part of opening a TCP connection.

File: Sources/PostgresNIO/Connection/PostgresConnection.swift [hunk 6]
```diff
@@ -68,31 +68,209 @@ public final class PostgresConnection {
...
+            /// - Parameters:
+            ///   - host: The hostname to connect to.
+            ///   - port: The TCP port to connect to (defaults to 5432).
+            ///   - connectTimeout: See ``connectTimeout``.
+            ///   - requireBackendKeyData: See ``requireBackendKeyData``.
+            public static func tcp(
+                host: String,
+                port: Int = 5432,
+                connectTimeout: TimeAmount = .seconds(10),
+                requireBackendKeyData: Bool = true
```
Comment (fabianfett) [inline Sources/PostgresNIO/Connection/PostgresConnection.swift:102]: as discussed yesterday, we should drop the tcp prefix where possible.

**Review by fabianfett — CHANGES_REQUESTED** [review]

File: Sources/PostgresNIO/Connection/PostgresConnection+Configuration.swift [hunk 7]
```diff
@@ -0,0 +1,318 @@
...
+            }
+            let base: Base
+            init(base: Base, tls: TLS) { (self.base, self.tls) = (base, tls) }
+        }
+        
+        // MARK: - Authentication
+
+        /// Contains authentication information for a connection.
+        public struct Authentication {
+            // TODO: Make the properties of this structure immutable (removing setters breaks public API).
```
Comment (fabianfett) [inline Sources/PostgresNIO/Connection/PostgresConnection+Configuration.swift:159]: Why do we want to do that?

File: Sources/PostgresNIO/Connection/PostgresConnection+Configuration.swift [hunk 8]
```diff
@@ -0,0 +1,318 @@
...
+            /// - Warning: This name is not validated in any way; in particular, there is no attempt to check that it
+            ///   matches the server's TLS certificate, nor is it in any way impacted by the validation setting in the
+            ///   TLS configuration (see ``PostgresConnection/Configuration/Server-swift.struct/TLS-swift.struct``).
+            public let tlsServerName: String?
+            
+            /// Whether the connection is required to provide backend key data (internal Postgres stuff).
+            ///
+            /// This property is provided for compatibility with Amazon RDS Proxy, which requires it to be `false`.
+            /// If you are not using Amazon RDS Proxy, you should probably leave this set to `true` (the default).
+            public let requireBackendKeyData: Bool
```
Comment (fabianfett) [inline Sources/PostgresNIO/Connection/PostgresConnection+Configuration.swift:209]: can we make the properties mutable?

File: Sources/PostgresNIO/Connection/PostgresConnection+Configuration.swift [hunk 9]
```diff
@@ -0,0 +1,318 @@
...
+            ///   - requireBackendKeyData: See ``requireBackendKeyData``. Default is `true`.
+            public init(
+                connectTimeout: TimeAmount = .seconds(10),
+                tlsServerName: String? = nil,
+                requireBackendKeyData: Bool = true
+            ) {
+                self.connectTimeout = connectTimeout
+                self.tlsServerName = tlsServerName
+                self.requireBackendKeyData = requireBackendKeyData
+            }
```
Comment (fabianfett) [inline Sources/PostgresNIO/Connection/PostgresConnection+Configuration.swift:227]: can we have just an init() that sets the default values? If users want to change this they can use the properties. This makes adding more properties to this way simpler in the future.

File: Sources/PostgresNIO/Connection/PostgresConnection+Configuration.swift [hunk 10]
```diff
@@ -0,0 +1,318 @@
...
+        ///   - options: See ``Options-swift.struct``. Most users do not need to adjust any of these.
+        public init(
+            server: Server,
+            authentication: Authentication,
+            options: Options = .init()
+        ) {
+            self.server = server
+            self.authentication = authentication
+            self.options = options
+        }
```
Comment (fabianfett) [inline Sources/PostgresNIO/Connection/PostgresConnection+Configuration.swift:259]: In an ideal case this is just an init and users do all changes through setting properties. See TLSConfiguration for comparison.

**Review by fabianfett — COMMENTED** [review]
Looks way better.

File: Sources/PostgresNIO/Connection/PostgresConnection.swift [hunk 11]
```diff
@@ -170,22 +68,15 @@ public final class PostgresConnection: @unchecked Sendable {
...
-                )
-                try channel.pipeline.syncOperations.addHandler(sslHandler, position: .first)
-            }
-        }
+            let sslHandler = try NIOSSLClientHandler(
+                context: sslContext,
+                serverHostname: configuration.serverNameForSNI
+            )
+            try channel.pipeline.syncOperations.addHandler(sslHandler, position: .first)
+        } }
```
Comment (fabianfett) [inline Sources/PostgresNIO/Connection/PostgresConnection.swift:79]: can we keep the impl here as is? I actually like that we use the internal enum here. This will make us think about it, if we add another case in the future. Also I find the create closure inside a closure extremely hard to read.

File: Sources/PostgresNIO/Connection/PostgresConnection+Configuration.swift [hunk 12]
```diff
@@ -0,0 +1,280 @@
...
+        case .configureChannel(let channel): self.connection = .bootstrapped(channel: channel)
+        }
+        self.username = config.username
+        self.password = config.password
+        self.database = config.database
+        self.tls = config.tls
+        self.options = config.options
+    }
+    
+    var serverNameForSNI: String? {
```
Comment (fabianfett) [inline Sources/PostgresNIO/Connection/PostgresConnection+Configuration.swift:257]: can we rename this to `serverNameForTLS`? Since it also used for certificate validation?

File: Sources/PostgresNIO/Connection/PostgresConnection.swift [hunk 13]
```diff
@@ -1,11 +1,11 @@
 import Atomics
 import NIOCore
+@_implementationOnly import NIOPosix
```
Comment (fabianfett) [inline Sources/PostgresNIO/Connection/PostgresConnection.swift:3]: why did you add `@_implementationOnly`?

File: Tests/PostgresNIOTests/New/PostgresChannelHandlerTests.swift [hunk 14]
```diff
@@ -36,9 +36,8 @@ class PostgresChannelHandlerTests: XCTestCase {
         XCTAssertNoThrow(try embedded.writeInbound(PostgresBackendMessage.readyForQuery(.idle)))
     }
     
-    func testEstablishSSLCallbackIsCalledIfSSLIsSupported() {
-        var config = self.testConnectionConfiguration()
-        XCTAssertNoThrow(config.tls = .require(try NIOSSLContext(configuration: .makeClientConfiguration())))
+    func testEstablishSSLCallbackIsCalledIfSSLIsSupported() throws {
+        let config = self.testConnectionConfiguration(tls: .require(try NIOSSLContext(configuration: .makeClientConfiguration())))
```
Comment (fabianfett) [inline Tests/PostgresNIOTests/New/PostgresChannelHandlerTests.swift:40]: can we keep the nothrow behavior here?

File: Tests/PostgresNIOTests/New/PostgresChannelHandlerTests.swift [hunk 15]
```diff
@@ -50,15 +49,15 @@ class PostgresChannelHandlerTests: XCTestCase {
         ])
         
         var maybeMessage: PostgresFrontendMessage?
-        XCTAssertNoThrow(embedded.connect(to: try .init(ipAddress: "0.0.0.0", port: 5432), promise: nil))
-        XCTAssertNoThrow(maybeMessage = try embedded.readOutbound(as: PostgresFrontendMessage.self))
+        embedded.connect(to: try .init(ipAddress: "0.0.0.0", port: 5432), promise: nil)
+        maybeMessage = try embedded.readOutbound(as: PostgresFrontendMessage.self)
```
Comment (fabianfett) [inline Tests/PostgresNIOTests/New/PostgresChannelHandlerTests.swift:53]: keep nothrow behavior

**Review by fabianfett — COMMENTED** [review]
LOVELY! Nits

**Review by fabianfett — APPROVED** [review]
Lovely! Thanks so much!
